web/handlers.go

```diff
@@ -190,13 +191,26 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if len(token) != 0 {
 		session, err := c.App.GetSession(token)
 		if err != nil {
-			c.Log.Info("Invalid session", mlog.Err(err))
-			if err.StatusCode == http.StatusInternalServerError {
-				c.Err = err
-			} else if h.RequireSession {
-				c.RemoveSessionCookie(w, r)
-				c.Err = model.NewAppError("ServeHTTP", "api.context.session_expired.app_error", nil, "token="+token, http.StatusUnauthorized)
+			if c.App.Srv().License() != nil && *c.App.Srv().License().Features.Cloud && tokenLocation == app.TokenLocationCloudHeader {
+				// Check to see if this provided token matches our CWS Token
+				session, err = c.App.GetCloudSession(token)
+				if err != nil {
+					c.Log.Info("Invalid CWS token", mlog.Err(err))
+					c.Err = err
+				} else {
+					c.App.SetSession(session)
+				}
```

Comment:
I wonder if we can move this specific case of creating a session into the `App.GetSession` and not act on an error case here

Reply:
Yes, probable makes sense to try to include that in the app package

Reply:
I've changed up this code to be a little less weird, you can see it in the latest diff, I think it's much cleaner than the first go.

I am hesitant to include this logic inside the `App.GetSession` function because that function has no notion of where the token came from, so conditionals will be challenging. We could pass the `tokenLocation` variable into `GetSession` to allow the logic to live entirely in that function. However, we'd then be changing the signature of a function used 17 times inside mattermost-server, and potentially more in plugins. That sounds like a dangerous change I'd like to avoid doing, especially in such a fundamental-to-the-app function like `GetSession`

Reply:
Looks much cleaner with the latest changes and +1 on changing the `App.GetSession` signature`